chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -1063,90 +1120,84 @@ impl PeerManagerActor {
         });
     }
 
-    /// Select one peer and send signal to stop connection to it gracefully.
-    /// Selection process:
-    ///     Create a safe set of peers, and among the remaining peers select one at random.
-    ///     If the number of outbound connections is less or equal than minimum_outbound_connections,
-    ///         add all outbound connections to the safe set.
-    ///     While the length of the safe set is less than safe_set_size:
-    ///         Among all the peers we have received a message within the last peer_recent_time_window,
-    ///             find the one we connected earlier and add it to the safe set.
-    ///         else break
-    fn try_stop_active_connection(&self) {
+    /// Check if the number of connections (excluding whitelisted ones) exceeds ideal_connections_hi.
+    /// If so, constructs a safe set of peers and selects one random peer outside of that set
+    /// and sends signal to stop connection to it gracefully.
+    ///     
+    /// Safe set contruction process:
+    /// 1. Add all whitelisted peers to the safe set.
+    /// 2. If the number of outbound connections is less or equal than minimum_outbound_connections,
+    ///    add all outbound connections to the safe set.
+    /// 3. Find all peers who sent us a message within the last peer_recent_time_window,
+    ///    and add them one by one to the safe_set (starting from earliest connection time)
+    ///    until safe set has safe_set_size elements.
+    fn maybe_stop_active_connection(&self) {
         debug!(target: "network",
             connected_peers_len = self.connected_peers.len(),
             ideal_connections_hi = self.config.ideal_connections_hi,
             "Trying to stop an active connection.",
         );
+        let filter_peers = |predicate: &dyn Fn(&ConnectedPeer) -> bool| {
+            self.connected_peers
+                .iter()
+                .filter_map(|(id, peer)| predicate(peer).then(|| id))
+                .collect::<Vec<_>>()
+        };
 
         // Build safe set
         let mut safe_set = HashSet::new();
 
-        if (self.connected_peers.values())
-            .filter(|connected_peer| connected_peer.peer_type == PeerType::Outbound)
-            .count()
-            + self.outgoing_peers.len()
-            <= self.config.minimum_outbound_peers as usize
+        // If there is not enough non-whitelisted peers, return without disconnecting anyone.
+        let whitelisted_peers =
+            filter_peers(&|p| self.is_peer_whitelisted(&p.full_peer_info.peer_info));
```

Comment:
are you sure? filter_peers takes a reference as an argument, because when I put just "dyn Fn(...)..." it complains that the closure is not Size. Is it possible to define generic closures?

Reply:
Oh, yes, you’re right.  I guess the closure would need to be replaced
by helper method:

```rust
fn filtered_connected_peer_ids(&self, predicate: impl Fn(info: &PeerInfo) -> bool) -> Vec<PeerId> {
    ...
}
```